leaderelection/leaderelection.go

```diff
@@ -0,0 +1,184 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Config is put into a different package to prevent cyclic imports in case
+// it is needed in several locations
+
+package leaderelection
+
+import (
+	"context"
+	"fmt"
+	"github.com/elastic/cloudbeat/config"
+	"github.com/elastic/cloudbeat/resources/providers"
+	"github.com/elastic/elastic-agent-autodiscover/kubernetes"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/gofrs/uuid"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s "k8s.io/client-go/kubernetes"
+	le "k8s.io/client-go/tools/leaderelection"
+	rl "k8s.io/client-go/tools/leaderelection/resourcelock"
+	"os"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	initOnce sync.Once
+	callOnce sync.Once
+	manager  *Manager
+)
+
+type ElectionManager interface {
+	IsLeader() bool
+	Run() error
+}
+
+type Manager struct {
+	log    *logp.Logger
+	ctx    context.Context
+	leader K8SLeaderElectionService
+	block  chan bool
+}
+
+// NewLeaderElector acts as a singleton
+func NewLeaderElector(ctx context.Context, log *logp.Logger, cfg config.Config) (ElectionManager, error) {
+	kubeClient, err := providers.KubernetesProvider{}.GetClient(cfg.KubeConfig, kubernetes.KubeClientOptions{})
+	if err != nil {
+		log.Errorf("NewLeaderElector error in GetClient: %v", err)
+		return nil, err
+	}
+
+	initOnce.Do(func() {
+		var leConfig le.LeaderElectionConfig
+		var leader K8SLeaderElectionService
+
+		block := make(chan bool)
+		leConfig, err = buildConfig(ctx, log, kubeClient, block)
+		leader, err = NewK8sLeaderElector(leConfig)
```

Comment:
a bit confusing because of the abstraction layer... not sure it's worth it if we're not using it

Reply:
yeah, pretty sure the best approach for now will be to get rid of it